src/threading/TaskScheduler.cpp

```diff
@@ -18,62 +19,231 @@
+        while (running.load() || !queue.at(pool.pool_id).empty()) {
+            // Wait at a high (but not realtime) priority to reduce latency
+            // for picking up a new task
+            update_current_thread_priority(1000);
+
+            run_task(get_task(pool.pool_id));
+        }
+    }
+
+    TaskScheduler::TaskScheduler() {
+        // Setup everything (thread pool, task queue, mutex, condition variable) for the main thread here
+        pools[util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID] =
+            util::ThreadPoolDescriptor{util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, 1};
+        pool_map[std::this_thread::get_id()] = util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID;
+        queue.emplace(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID, std::vector<std::unique_ptr<ReactionTask>>{});
+    }
+
+    void TaskScheduler::start_threads(const util::ThreadPoolDescriptor& pool) {
+        // The main thread never needs to be started
+        if (pool.pool_id != util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID) {
+            const std::lock_guard<std::mutex> pool_lock(pool_mutex);
+            for (size_t i = 0; i < pool.thread_count; ++i) {
+                threads.push_back(std::make_unique<std::thread>(&TaskScheduler::pool_func, this, pool));
+                pool_map[threads.back()->get_id()] = pool.pool_id;
+            }
+        }
+    }
+
+    void TaskScheduler::create_pool(const util::ThreadPoolDescriptor& pool) {
+        // Pool already exists
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> pool_lock(pool_mutex);
+            if (pools.count(pool.pool_id) > 0 && pools.at(pool.pool_id).thread_count > 0) {
+                return;
+            }
+
+            // Make a copy of the pool descriptor
+            pools[pool.pool_id] = util::ThreadPoolDescriptor{pool.pool_id, pool.thread_count};
+        }
+
+        // Make sure the task queue is created for this pool
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+            if (queue.count(pool.pool_id) == 0) {
+                queue.emplace(pool.pool_id, std::vector<std::unique_ptr<ReactionTask>>{});
+            }
+        }
+
+        // If the scheduler has not yet started then don't start the threads for this pool yet
+        if (started.load()) {
+            start_threads(pool);
         }
-        condition.notify_all();
+    }
+
+    void TaskScheduler::start(const size_t& thread_count) {
+
+        // Make the default pool
+        create_pool(util::ThreadPoolDescriptor{util::ThreadPoolDescriptor::DEFAULT_THREAD_POOL_ID, thread_count});
+
+        // The scheduler is now started
+        started.store(true);
+
+        // Start all our threads
+        for (const auto& pool : pools) {
+            start_threads(pool.second);
+        }
+
+        // Run main thread tasks
+        pool_func(pools.at(util::ThreadPoolDescriptor::MAIN_THREAD_POOL_ID));
+
```

Comment:
I would put a comment here to say that when the main thread reaches this point it's because the powerplant has shutdown and the main thread is leaving the task scheduler